packages/@aws-cdk/aws-rds/lib/cluster-engine.ts

```diff
@@ -0,0 +1,286 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
+import * as core from '@aws-cdk/core';
+import { ClusterParameterGroup, IParameterGroup, ParameterGroup } from './parameter-group';
+import { ParameterGroupFamilyMapping } from './private/parameter-group-family-mapping';
+import { compare } from './private/version';
+
+/**
+ * The extra options passed to the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineBindOptions {
+  /**
+   * The role used for S3 importing.
+   *
+   * @default - none
+   */
+  readonly s3ImportRole?: iam.IRole;
+
+  /**
+   * The role used for S3 exporting.
+   *
+   *  @default - none
+   */
+  readonly s3ExportRole?: iam.IRole;
+
+  /**
+   * The customer-provided ParameterGroup.
+   *
+   * @default - none
+   */
+  readonly parameterGroup?: IParameterGroup;
+}
+
+/**
+ * The type returned from the {@link IClusterEngine.bindToCluster} method.
+ */
+export interface ClusterEngineBindConfig {
+  /**
+   * The ParameterGroup to use for the cluster.
+   *
+   * @default - no ParameterGroup will be used
+   */
+  readonly parameterGroup?: IParameterGroup;
+
+  /**
+   * The port to use for this cluster,
+   * unless the customer specified the port directly.
+   *
+   * @default - use the default port for clusters (3306)
+   */
+  readonly port?: number;
+}
+
+/**
+ * The interface representing a database cluster (as opposed to instance) engine.
+ */
+export interface IClusterEngine {
+  /** The type of the engine, for example "aurora-mysql". */
+  readonly engineType: string;
```

Comment:
Why the dichotomy between what's returned via `bind()` and what's immediately available on the cluster engine?

Taking a guess that it's born from necessity from what we need *right now*, but I don't like this pattern much. I'm worried it won't grow well with changing requirements in the future. How sure are we that some of the instance properties won't become `bind()`-dependent in the future?

Reply:
Yes, there is a dichotomy. The reason is that the things are useful at different stages of the lifecycle of other objects, without being bound to a Cluster/Instance.

For example, the property `parameterGroupFamily` is useful when creating `ParameterGroup`s:

```ts
new rds.ClusterParameterGroup(this, 'PG', {
  family: rds.DatabaseClusterEngine.AURORA.parameterGroupFamily,
  // ...
```

(that's why I suggested we should maybe make `ClusterParameterGroup` take an `IClusterEngine` in its construction parameters)

The `defaultInstanceType` is useful when creating Instances, for example:

```ts
new rds.DatabaseInstanceReadReplica(stack, 'ReadReplica', {
  instanceType: rds.DatabaseInstanceEngine.POSTGRES.defaultInstanceType,
  // ...
```

This isn't very different than what we do for many other things that have a `bind` method (example: [CodeBuild `IArtifacts`](https://github.com/aws/aws-cdk/blob/f6fe36a0281a60ad65474b6ce0e22d0182ed2bea/packages/%40aws-cdk/aws-codebuild/lib/artifacts.ts#L20-L39)).